File: matlab/readsonarfile.m
% Example of reading a Tritech or ARIS sonar file in Matlab using the
% TritechFiles Java library from https://github.com/douggillespie/TritechFiles
% see the Javadoc at
% file:///C:/Users/dg50/source/repos/TritechFiles/docs/index.html for the
% API and function call details. The main classes you'll be wanting to use
% to read data are MultiFileCatalog to efficiently get access to all
% records in a file or folder of files, and SonarImageRecord and 
% it's device specific subclasses. 
% FanPicksFromData can be used to convert raw data to a fan image. 

% find the jar file, check it's path and add full path to Matlab java path
jarFile = 'C:\\Users\\dg50\\source\\repos\\TritechFiles\\TritechFilesV0102.jar';
% jarFile = which(jarFileName)
% not very important, but will get a waring if the jar is loaded a second time
warning('off') 
javaaddpath(jarFile);

% the MultiFileCatalog can work with a single file, or a folder of files.
% For this example, we'll just use a single file
sampleFile = "C:\ProjectData\Meygen2022\AAM\GLFData\Drive23\11-15\log_2022-11-15-181117.glf";

mfc = tritechgemini.fileio.MultiFileCatalog;
mfc.catalogFiles(sampleFile);

% see which sonars are present in the data
sonars = mfc.getSonarIDs;
% Make a fan maker for each sonar. It's not good to share them since they
% contain a lookup table that has to be recreated whenever the data
% dimension changes
for i = 1:numel(sonars)
    fanMaker{i} = tritechgemini.imagedata.FanPicksFromData(4);
end
% loop through the fist 6 records, plotting each
figure(1)
clf
figure(2)
clf
for i = 1:6
    aRec = mfc.getRecord(i-1);
    % get the rectangular image data.
    raw = aRec.getShortImageData; 
    % that comes in as a 1D array, so reshape and transpose
    raw = reshape(raw, aRec.getnBeam, aRec.getnRange)';
    figure(1)
    subplot(3,2,i)
    imagesc(raw)
    axis xy
    xlabel('Beam');
    ylabel('Range')

    for f = 1:numel(sonars) % find the fan maker
        if aRec.getDeviceId() == sonars(f)
            fm = fanMaker{f};
        end
    end
    % make the fan image
    fanImage = fm.createFanData(aRec, 1000);
    % get the values
    image = fanImage.getImageValues;
    image = image';  % tranpose
    % get the scales (should be the same)
    xS = fanImage.getMetresPerPixX;
    yS = fanImage.getMetresPerPixY;
    % make scales in meters to match the data. 
    xBins = ([1:size(image,2)]-size(image,2)/2)*xS;
    yBins = [1:size(image,1)]*yS;
    % plot the image and flip to correct axis
    figure(2);
    subplot(3,2,i)
    imagesc(xBins, yBins,image)
    axis xy
    axis equal
    xlabel('Metres');
    ylabel('(m)')
end

File: R/readsonarfile.R
# read a Tritech sonar data file and plot a frame in R
# install.packages('rJava')
library(rJava)
library(plot3D)

# initialise Java and add the jar library
.jinit()
jarFile <- 'C:\\Users\\dg50\\source\\repos\\TritechFiles\\TritechFilesV0102.jar';
.jaddClassPath(jarFile) 

sampleFile <- 'C:/ProjectData/Meygen2022/AAM/GLFData/Drive23/11-15/log_2022-11-15-181117.glf';
sampleFolder <- 'C:/ProjectData/Meygen2022/AAM/GLFData/Drive23/11-15/';
# Create the multifile catalog for the single file, or folder. 
mfc <- .jnew("tritechgemini.fileio.MultiFileCatalog")
# print out all the methods in the catalog class to see they are there 
.jmethods(mfc) 
# catalog a single file, or folder of files. 
b <- J(mfc, "catalogFiles", sampleFolder)
nRec <- J(mfc, "getTotalRecords")
print(paste('total number of records is', nRec))

# now create the FanPicksFromData object used to transform images.
fanMaker <- .jnew("tritechgemini.imagedata.FanPicksFromData", as.integer(4))

# read in the first record.If looping through records, you'd start your loop here 
aRec <- J(mfc, "getRecord", as.integer(0))
# get the raw data out
raw <- J(aRec, "getShortImageData")
nBeam <- J(aRec, "getnBeam")
nRange <- J(aRec, "getnRange")
# data come in as a 1D array. Reshape it to 2D. 
raw2 <- array(raw, dim=c(nBeam, nRange))
plot.new()
image2D(raw2, main="Raw image", xlab="Beam", ylab="Range")

# now make a fan image from the record we read in to be 500 pixels wide (quite small)
fanImage <- J(fanMaker, "createFanData", aRec, as.integer(500))
imageData <- J(fanImage, "getImageValues")
# now free the image data in the record, or you'll run out of memory
# after a few thousand records.
J(aRec, "freeImageData")
# imageData comes back as a 2D array of Java objects and needs to be converted
# This works as found at https://stackoverflow.com/questions/17556910/accessing-data-from-java-object-in-rjava
rImage <- do.call(rbind, lapply(imageData, .jevalArray))
# get the image dimensions
imageDim = dim(rImage);
# create the scales, should be the same in x and y
xScale <- J(fanImage, "getMetresPerPixX")
xs = (seq(1, imageDim[1])-imageDim[1]/2) *xScale
ys = seq(1, imageDim[2]) * xScale;
# plot it
plot.new()
image2D(rImage, x = xs, y = ys, main="Fan Image", xlab="(m)", ylab="(m)")
